docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,229 @@
+called at the beginning or after calling oe_debug_malloc_tracking_stop(). While
+the calling sequence following this rule, the return values are false. TRUE, or
+non-zero return value means the sequence is wrong and the report does not
+reflect the real situation.
+
+- oe_debug_malloc_tracking_start() and oe_debug_malloc_tracking_stop() can been
+called many times:
+
+```
+  |----code1--|-----------|----code2--|--|----code3--|
+  start1->stop1->report1->start2->stop2->start3->stop3->report3
+```
+
+In this case report1 only covers code1, and report3 covers code1, code2 and
+code3 cumulatively.
+
+# Implementation
+
+The debug malloc employees a linkedlist to make record of all memory operations.
+So, for the local detection, we can make use of the fulfilled record system.
+
+```c
+struct header
+{
+    /* Contains HEADER_MAGIC1 */
+    uint64_t magic1;
+
+    /* Headers are kept on a doubly-linked list */
+    header_t* next;
+    header_t* prev;
+
+    /* The alignment passed to memalign() or zero */
+    uint64_t alignment;
+
+    /* Size of user memory */
+    size_t size;
+
+    /* Return addresses obtained by oe_backtrace() */
+    void* addrs[OE_BACKTRACE_MAX];
+    uint64_t num_addrs;
+
+    /* Option if current object is tracked */
+    bool local_tracking;
+
+    /* Padding to make header a multiple of 16 */
+    uint8_t padding[7];
+
+    /* Contains HEADER_MAGIC2 */
+    uint64_t magic2;
+
+    /* User data */
+    uint8_t data[];
+};
+```
+
+We also add an variable
+```c
+bool oe_use_debug_malloc_tracking = false;
+```
+which will be changed by oe_debug_malloc_tracking_start() and
+oe_debug_malloc_tracking_stop(). While doing the allocation, the objective's
+`local_tracking` is marked as `oe_use_debug_malloc_tracking`, then
+oe_debug_malloc_tracking_report() can get the number of all the not yet freed
+objects.
+
+# Sample
+
+```c
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/debugmalloc.h>
+#include <stdio.h>
+
+// Include the trusted debugmalloc header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the debugmalloc.edl file.
+#include "debugmalloc_t.h"
+
+// This is the function that the host calls. It prints
+// a message in the enclave before calling back out to
+// the host to print a message from there too.
+void enclave_hello()
+{
+    void* buf1, buf2, buf3;
+    int size1, size2;
+
+    oe_debug_malloc_tracking_start();
+    void* buf1 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+    size1 = oe_debug_malloc_tracking_report();
+
+    oe_debug_malloc_tracking_start();
+    void* buf2 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+
+    oe_debug_malloc_tracking_start();
+    void* buf3 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+    size2 = oe_debug_malloc_tracking_report();
+
+
+    printf("There are %d, %d un-freed objects reported", size1, size2);
```

Comment:
This line is confusing.  Put a separate printf for size1 up after line 186 and keep this one just about size2.

Reply:
Updated, thanks!